src/Streamly/Internal/Data/Parser/Types.hs

```diff
@@ -181,7 +252,220 @@ splitWith func (Parser stepL initialL extractL)
+            Error _ -> do
+                rR <- initialR
+                return $ Skip (cnt + 1) (AltParseR rR)
+
+    step (AltParseR st) a = do
+        r <- stepR st a
+        return $ case r of
+            Yield n s -> Yield n (AltParseR s)
+            Skip n s -> Skip n (AltParseR s)
+            Stop n b -> Stop n b
+            Error err -> Error err
+
+    extract (AltParseR sR) = extractR sR
+    extract (AltParseL _ sL) = extractL sL
+
+-- | See documentation of 'Streamly.Internal.Data.Parser.many'.
+--
+-- /Internal/
+--
+{-# INLINE splitMany #-}
+splitMany :: MonadCatch m => Fold m b c -> Parser m a b -> Parser m a c
+splitMany (Fold fstep finitial fextract) (Parser step1 initial1 extract1) =
+    Parser step initial extract
+
+    where
+
+    initial = do
+        ps <- initial1 -- parse state
+        fs <- finitial -- fold state
+        pure (Tuple3' ps 0 fs)
+
+    {-# INLINE step #-}
+    step (Tuple3' st cnt fs) a = do
+        r <- step1 st a
+        let cnt1 = cnt + 1
+        case r of
+            Yield _ s -> return $ Skip 0 (Tuple3' s cnt1 fs)
+            Skip n s -> do
+                assert (cnt1 - n >= 0) (return ())
+                return $ Skip n (Tuple3' s (cnt1 - n) fs)
+            Stop n b -> do
+                s <- initial1
+                fs1 <- fstep fs b
+                -- XXX we need to yield and backtrack here
+                return $ Skip n (Tuple3' s 0 fs1)
+            Error _ -> do
+                xs <- fextract fs
+                return $ Stop cnt1 xs
+
+    -- XXX The "try" may impact performance if this parser is used as a scan
+    extract (Tuple3' s _ fs) = do
+        r <- try $ extract1 s
+        case r of
+            Left (_ :: ParseError) -> fextract fs
+            Right b -> fstep fs b >>= fextract
+
+-- | See documentation of 'Streamly.Internal.Data.Parser.some'.
+--
+-- /Internal/
+--
+{-# INLINE splitSome #-}
+splitSome :: MonadCatch m => Fold m b c -> Parser m a b -> Parser m a c
+splitSome (Fold fstep finitial fextract) (Parser step1 initial1 extract1) =
+    Parser step initial extract
+
+    where
+
+    initial = do
+        ps <- initial1 -- parse state
+        fs <- finitial -- fold state
+        pure (Tuple3' ps 0 (Left fs))
+
+    {-# INLINE step #-}
+    step (Tuple3' st _ (Left fs)) a = do
+        r <- step1 st a
+        case r of
+            Yield _ s -> return $ Skip 0 (Tuple3' s undefined (Left fs))
+            Skip  n s -> return $ Skip n (Tuple3' s undefined (Left fs))
+            Stop n b -> do
+                s <- initial1
+                fs1 <- fstep fs b
+                -- XXX this is also a yield point, we will never fail beyond
+                -- this point. If we do not yield then if an error occurs after
+                -- this then we will backtrack to the previous yield point
+                -- instead of this point which is wrong.
+                --
+                -- so we need a yield with backtrack
```

Comment:
Is this intentionally left out?

Reply:
Will fix it later in a separate change. It will require changes everywhere.